### ISSUE TITLE
InvalidOperationException - No reading operation to complete

### ISSUE THREAD

**Issue by Michael-Kempe** [StephenClearyExamples/TcpChat#1]
If I send 2 messages in a row from the client, the serer got an InvalidOperationException (No reading operation to complete).

Here is my clients Send-Click-Event Code that I have changed - see 2nd call to SendMessageAsync:
```

private async void Button_Click_1(object sender, RoutedEventArgs e)
{
	if (_chatConnection == null)
	{
		Log.Text += "No connection!\n";
	}
	else
	{
		await _chatConnection.SendMessageAsync(new ChatMessage(chatMessageTextBox.Text));
		// this is my additional SendMessageAsync call
		await _chatConnection.SendMessageAsync(new ChatMessage(chatMessageTextBox.Text));
		Log.Text += $"Sent message: {chatMessageTextBox.Text}\n";
	}
}

```

The server exception is raised in ProcessSocketMainLoop.

![image](https://user-images.githubusercontent.com/46563088/157604282-2d0378dd-d1d3-4939-8525-304d3ab3c79d.png)

Any hints on this?

**Comment by fredimachado** [comment]
Hey @Michael-Kempe!

I think I found the issue. But I would wait for @StephenCleary to be sure, as I might not be aware of some edge cases, so maybe there's a more proper way to fix this.

https://github.com/fredimachado/TcpChat/commit/d550ecfd89159fdf327ff46d5e110eea38557319

**Comment by Michael-Kempe** [comment]
Ah, I see.  This fix my problem! Cool. But like you said... let's wait for @StephenCleary

**Comment by fredimachado** [comment]
Yeah, in case `sequenceReader.TryReadMessage` returns false I think my change will probably make things blow up haha.

**Comment by StephenCleary** [comment]
Thanks for finding this!

So, the problem was that I was calling `AdvanceTo` once for each message, and that was wrong. There's this ping/pong style of API that the pipelines use: get a buffer, then advance, then get another buffer, then advance, etc. So advancing twice on the same buffer was my bad.

Fixed in https://github.com/StephenClearyExamples/TcpChat/commit/f4c8b1c5564ccc0a50f04f11fc93409df4d0e303

I'm going to leave this issue open until I find where I introduced the bug and add a note in the video description about it.
